hybrid/detailed_pv_plant.py

```diff
@@ -0,0 +1,103 @@
+from typing import Union, Optional, Sequence, Any
+
+import PySAM.Pvsamv1 as Pvsam
+import PySAM.Singleowner as Singleowner
+
+from hybrid.power_source import *
+from hybrid.layout.pv_design_utils import *
+from hybrid.layout.pv_layout import PVLayout, PVGridParameters
+from hybrid.dispatch.power_sources.pv_dispatch import PvDispatch
+
+
+class DetailedPVPlant(PowerSource):
+    _system_model: Pvsam.Pvsamv1
+    _financial_model: Union[Any, Singleowner.Singleowner]
+    _layout: Union[Any, PVLayout]
+    _dispatch: PvDispatch
+
+    def __init__(self,
+                 site: SiteInfo,
+                 pv_config: dict):
+        """
+
+        :param pv_config: dict, with following keys:
+            'tech_config': dict, contains parameters for pvsamv1 technology model
+            'fin_config': dict, contains `model_type` and any inputs for chosen financial model type
+            'layout_params': optional DetailedPVParameters, the design vector w/ values. Required for layout modeling
+            'layout_config': optional dict, contains all keys for PVLayoutConfig dataclass. Required for layout modeling
+        """
+        if 'tech_config' not in pv_config.keys():
+            raise ValueError
+
+        system_model = Pvsam.default("FlatPlatePVSingleOwner")
+        financial_model = Singleowner.from_existing(system_model, "FlatPlatePVSingleOwner")
+
+        super().__init__("SolarPlant", site, system_model, financial_model)
+
+        self._system_model.SolarResource.solar_resource_data = self.site.solar_resource.data
+
+        self.dc_degradation = [0]
+
+        params: Optional[PVGridParameters] = None
+        if 'layout_params' in pv_config.keys():
+            params: PVGridParameters = pv_config['layout_params']
+        self._layout = PVLayout(site, system_model, params)
+
+        self._dispatch: PvDispatch = None
+
+        self.processed_assign(pv_config['tech_config'])
+
+    def processed_assign(self, params):
+        """
+        Assign attributes from dictionary with additional processing
+        to enforce coherence between attributes
+        """
+        self.assign(params)
+        n_strings, system_capacity, n_inverters = align_from_capacity(
+            system_capacity_target=self.value('system_capacity'),
+            modules_per_string=self.value('subarray1_modules_per_string'),
+            module_power=get_module_power(self._system_model) * 1e-3,
+            inverter_power=get_inverter_attribs(self._system_model)['P_ac'] * 1e-3,
+            n_inverters_orig=self.value('inverter_count')
+        )
+        self._system_model.SystemDesign.subarray1_nstrings = n_strings
+        self._system_model.SystemDesign.system_capacity = system_capacity
+        self._system_model.SystemDesign.inverter_count = n_inverters
```

Comment:
How I've tried to handle this issue of interdependent variables is shown in the wind source. For instance, in the `__init__`, the turbine rating is assigned from the config, and the rating then changes the system capacity:

https://github.com/NREL/HOPP/blob/c4e2a55976d18f3fd8f222604979035181448174/hybrid/wind_source.py#L71

The `turb_rating` is actually a Python class property with custom-defined getters and setters, as you can see here:

https://github.com/NREL/HOPP/blob/c4e2a55976d18f3fd8f222604979035181448174/hybrid/wind_source.py#L130

Whenever the setter `self.turb_rating = x` is called, we end up going to that custom function. That custom function's job is to make everything in PySAM consistent to the best of its ability.

When you have these new Pvsamv1 variables that users are supposed to use, you will need to add these class properties to control how they are accessed and modified. We want to enable the ability for the user to do:

```
pvplant.n_strings = 11
pvplant.module_power = 240
pvplant.subarray1_modules_per_string = 10
...
```

and have the end resulting Pvsamv1 model consistent.

Reply:
I've been running into trouble with the way some properties are set up in HOPP, where you get caught in loops, and that's why I'm setting these directly. In this case, the _actual_ system_capacity has been calculated (without designing a new layout) and should be set in the pysam model as-is, where the property is set up to assume it's been given the _target_ system_capacity and it then calculates a new layout and sets a different system_capacity.